soccer/joystick/ManualControlNode.cpp

```diff
@@ -0,0 +1,246 @@
+#include "ManualControlNode.hpp"
+
+namespace joystick {
+REGISTER_CONFIGURABLE(ManualControlNode);
+
+ConfigDouble* ManualControlNode::JoystickRotationMaxSpeed;
+ConfigDouble* ManualControlNode::JoystickRotationMaxDampedSpeed;
+ConfigDouble* ManualControlNode::JoystickTranslationMaxSpeed;
+ConfigDouble* ManualControlNode::JoystickTranslationMaxDampedSpeed;
+
+ManualControlNode::ManualControlNode(Context* context) : context_{context} {}
+
+void ManualControlNode::run() {
+    // Update list of connected gamepads
+    updateGamepadList();
+
+    // "Manually" call the fake callback on the data in context_
+    for (const auto& msg : context_->gamepad_messages) {
+        callback(msg);
+    }
+
+    const auto robots = context_->state.self;
+    // Set all robots to not be joystick controlled
+    for (OurRobot* robot : robots) {
+        robot->setJoystickControlled(false);
+    }
+
+    // If there aren't any gamepads that are connected, return
+    if (gamepad_stack_.empty()) {
+        return;
+    }
+
+    const int manual_id = context_->game_settings.joystick_config.manualID;
+
+    // Find the robot that we are supposed to control
+    const auto pred = [manual_id](const OurRobot* r) {
+        return r->shell() == manual_id;
+    };
+    const auto it = std::find_if(robots.begin(), robots.end(), pred);
+
+    // Return if we can't find the robot we're supposed to control
+    if (it == robots.end()) {
+        return;
+    }
+
+    OurRobot* robot = *it;
+
+    robot->setJoystickControlled(true);
+
+    updateIntentAndSetpoint(robot);
+}
+
+void ManualControlNode::updateIntentAndSetpoint(OurRobot* robot) {
+    int shell = robot->shell();
+
+    // Modify robot->setpoint and robot->intent
+    RobotIntent& intent = context_->robot_intents[shell];
+    MotionSetpoint& setpoint = context_->motion_setpoints[shell];
+
+    intent.is_active = true;
+
+    float x_vel = controls_.x_vel;
+    float y_vel = controls_.y_vel;
+
+    // Field oriented control
+    if (robot->visible() &&
+        context_->game_settings.joystick_config.useFieldOrientedDrive) {
+        Geometry2d::Point translation{x_vel, y_vel};
+        translation.rotate(-M_PI / 2 - robot->angle());
+
+        x_vel = static_cast<float>(translation.x());
+        y_vel = static_cast<float>(translation.y());
+    }
```

Comment:
If FOC is on and the bot is not visible I would just stop it.

Reply:
done